Rakefile

```diff
@@ -1,66 +1,112 @@
+PROJ_DIR      = Pathname.new(__dir__)
+APPS_DIR      = PROJ_DIR.join('apps')
+GEMFILE       = PROJ_DIR.join('Gemfile')
+INSTALL_ROOT  = Pathname.new(ENV["PREFIX"] || "/opt/ood")
+VENDOR_BUNDLE = (ENV['VENDOR_BUNDLE'] == "yes" || ENV['VENDOR_BUNDLE'] == "true")
 
-def all_components
-  JSON.parse(CONFIG_FILE.read).map { |c| Component.new(c) }
+def apps
+  Dir["#{APPS_DIR}/*"].map { |d| Component.new(d) }
 end
 
 class Component
   attr_reader :name
-  attr_reader :url
-  attr_reader :tag
-
-  def initialize(opts = {})
-    @name = opts.fetch("name") { raise ArgumentError, "No name specified. Missing argument: name" }.to_s
-    @url  = opts.fetch("url")  { raise ArgumentError, "No url specified. Missing argument: url" }.to_s
-    @tag  = opts.fetch("tag")  { raise ArgumentError, "No tag specified. Missing argument: tag" }.to_s
-    @app  = opts.fetch("app", nil)
+  attr_reader :path
+
+  def initialize(app)
+    @name = File.basename(app)
+    @path = Pathname.new(app)
   end
 
-  def app?
-    !!@app
+  def ruby_app?
+    @path.join('config.ru').exist?
   end
 
-  def build_root
-    app? ? BUILD_ROOT.join("apps", name) : BUILD_ROOT.join(name)
+  def node_app?
+    @path.join('app.js').exist?
   end
+end
 
-  def download_url
-    "#{url}/archive/#{tag}.tar.gz"
+desc "Package OnDemand"
+task :package do
+  `which gtar 1>/dev/null 2>&1`
+  if $?.success?
+    tar = 'gtar'
+  else
+    tar = 'tar'
   end
+  version = ENV['VERSION']
+  if ! version
+    latest_commit = `git rev-list --tags --max-count=1`.strip[0..6]
+    latest_tag = `git describe --tags #{latest_commit}`.strip[1..-1]
+    datetime = Time.now.strftime("%Y%m%d-%H%M")
+    version = "#{latest_tag}-#{datetime}-#{latest_commit}"
+  end
+  sh "git ls-files | #{tar} -c --transform 's,^,ondemand-#{version}/,' -T - | gzip > packaging/v#{version}.tar.gz"
 end
 
-task :default => :build
+namespace :build do
+  desc "Build gems"
+  task :gems do
+    if VENDOR_BUNDLE
+      args = "--path vendor/bundle"
+    else
+      args = ""
+    end
+    apps.each do |a|
+      next unless a.ruby_app?
+      chdir a.path do
+        sh "bundle install #{args}"
+      end
+    end
+  end
 
-all_components.each do |c|
-  file c.build_root => CONFIG_FILE do
-    rm_rf c.build_root if c.build_root.directory?
-    mkdir_p c.build_root unless c.build_root.directory?
-    sh "curl -skL #{c.download_url} | tar xzf - -C #{c.build_root} --strip-components=1"
-    setup_path = c.build_root.join("bin", "setup")
-    if setup_path.exist? && setup_path.executable?
-      sh "PASSENGER_APP_ENV=production PASSENGER_BASE_URI=/pun/sys/#{c.name} #{setup_path}"
+  apps.each do |a|
+    if a.ruby_app?
+      depends = [:gems]
+    else
+      depends = []
+    end
+    task a.name.to_sym => depends do |t|
+      setup_path = a.path.join("bin", "setup")
+      if setup_path.exist? && setup_path.executable?
+        sh "PASSENGER_APP_ENV=production PASSENGER_BASE_URI=/pun/sys/#{a.name} #{setup_path}"
```

Comment:
We can omit this:

```
PASSENGER_BASE_URI=/pun/sys/#{a.name}
```

The default of each app's bin/setup is to set the sys URL. i.e. https://github.com/OSC/ondemand/blob/a16f28858a2ca82dfd4868842c9cdacb7f5a65d4/apps/dashboard/bin/setup#L11

However, it is important to set `PASSENGER_APP_ENV=production` so keep that.

Reply:
Change made.